core/ledger.re

```diff
@@ -0,0 +1,176 @@
+  ticket_table: Ticket_table.empty,
+  handles: Handle_tree.empty,
+};
+
+let balance = (address, ticket, t) => {
+  let balance = {
+    let.some handle =
+      Address_and_ticket_map.find_opt(address, ticket, t.ledger);
+    let.some ticket = Ticket_table.find_opt(handle, t.ticket_table);
+    Some(ticket.amount);
+  };
+  balance |> Option.value(~default=Amount.zero);
+};
+
+let get_or_create = (address, ticket, ticket_table, ledger) => {
+  switch (Address_and_ticket_map.find_opt(address, ticket, ledger)) {
+  | None =>
+    let (handle, ticket_table) =
+      Ticket_table.add_empty(ticket, ticket_table);
+    let ledger = Address_and_ticket_map.add(address, ticket, handle, ledger);
+    (handle, ticket_table, ledger);
+  | Some(handle) => (handle, ticket_table, ledger)
+  };
+};
+
+let transfer = (~source, ~destination, amount, ticket, t) => {
+  let.ok source_handle =
+    Address_and_ticket_map.find_opt(source, ticket, t.ledger)
+    |> Option.to_result(~none=`Not_enough_funds);
+  let (destination_handle, ticket_table, ledger) =
+    get_or_create(destination, ticket, t.ticket_table, t.ledger);
+  let.ok (Ticket_table.{split_at, remaining: source_handle}, ticket_table) =
+    Ticket_table.split(source_handle, ~at=amount, ticket_table)
+    |> Result.map_error(
+         fun
+         | `Invalid_ticket => failwith("Should be impossible")
+         | `Not_enough_funds as e => e,
+       );
+
+  let (destination_handle, ticket_table) =
+    Ticket_table.join(destination_handle, split_at, ticket_table)
+    |> Result.get_ok;
+
+  Ok({
+    ledger:
+      ledger
+      |> Address_and_ticket_map.add(source, ticket, source_handle)
+      |> Address_and_ticket_map.add(destination, ticket, destination_handle),
+    ticket_table,
+    handles: t.handles,
+  });
+};
+
+// tezos operations
+let deposit = (destination, amount, ticket, t) => {
+  let (to_merge_handle, ticket_table) =
+    Ticket_table.add(
+      Ticket_table.unsafe_create_ticket(ticket, amount),
+      t.ticket_table,
+    );
+  let (destination_handle, ticket_table, ledger) =
+    get_or_create(destination, ticket, ticket_table, t.ledger);
+  let (destination_handle, ticket_table) =
+    Ticket_table.join(destination_handle, to_merge_handle, ticket_table)
+    |> Result.get_ok;
+  {
+    ledger:
+      ledger
+      |> Address_and_ticket_map.add(destination, ticket, destination_handle),
+    ticket_table,
+    handles: t.handles,
+  };
+};
+let withdraw = (~source, ~destination, amount, ticket, t) => {
+  let owner = destination;
+
+  let.ok source_handle =
+    Address_and_ticket_map.find_opt(source, ticket, t.ledger)
+    |> Option.to_result(~none=`Not_enough_funds);
+
+  let.ok (Ticket_table.{split_at, remaining}, ticket_table) =
+    Ticket_table.split(source_handle, ~at=amount, t.ticket_table)
+    |> Result.map_error(
+         fun
+         | `Invalid_ticket => failwith("Should be impossible")
+         | `Not_enough_funds as e => e,
+       );
```

Comment:
this is all over this module unfortunately and I'm not sure how to get rid of it, is this the right thing to do?